working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1501,17 +1538,44 @@ The context type schema for a pattern `p` is:
         doesn't destructure anything, it matches any list, so it is permissive
         with the context type.*
 
-    3.  Else `E` is the greatest lower bound of the type schemas of all element
-        subpatterns. *We use the greatest lower bound to ensure that the outer
-        collection type has a precise enough type to ensure that any typed field
-        subpatterns do not need to downcast:*
+    3.  Else, infer the type schema from the subpatterns:
 
-        ```dart
-        var [int a, num b] = [1, 2];
-        ```
+        1.  Let `es` be an empty list of types.
+
+        2.  For each subpattern `e` in `p`:
+
+            1.  If `e` is a rest element pattern with a subpattern `s` and the
+                context type schema of `s` is a `List<T>` for some type `T`,
+                then add `T` to `es`.
+
+            2.  Else if `e` is not a rest element pattern, add the context
+                type schema of `e` to `es`.
 
-        *Here, the GLB of `int` and `num` is `int`, which ensures that neither
-        `int a` nor `num b` need to downcast their respective fields.*
+            *Else, `e` is a rest element with no subpattern, so it doesn't
+            contribute to inference.*
+
+        3.  If `es` is empty, then `E` is `Object?`. *This can happen if the
```

Comment:
I don't have an opinion about whether `Object?` or the "unknown type" is a better choice, but whatever we do here, we should do the same for bullet point 2 on line 1537.

Or maybe it would be better to just drop that bullet point entirely.  (We previously needed it because the GLB of zero types was undefined.  But now that we're explicitly defining what to do if no types get accumulated into `es`, there's no need to separately handle the case of an empty list).

Reply:
Good point. Changed to `?` here and elsewhere (constant patterns, unknown maps, etc.).